eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1306 @@
+                                   mustache_token_section_t *token)
+{
+  g_autofree char *msg = NULL;
+
+  if (g_variant_is_of_type (variant, G_VARIANT_TYPE_STRING_ARRAY))
+    return _renderer_strv_sect_from_ht (api, userdata, token, variant);
+  else if (g_variant_is_of_type (variant, G_VARIANT_TYPE_BOOLEAN))
+    return _renderer_bool_sect_from_ht (api, userdata, token, variant);
+  else if (g_variant_is_of_type (variant, G_VARIANT_TYPE_STRING))
+    return _renderer_str_sect_from_ht (api, userdata, token, variant);
+
+  msg = g_strdup_printf ("No handler for section type %s on token %s",
+                         g_variant_get_type_string (variant),
+                         token->name);
+  (*api->error) (api, userdata, __LINE__, msg);
+  return 1;
+}
+
+static uintmax_t
+_renderer_sect_from_ht (mustache_api_t           *api,
+                        void                     *userdata,
+                        mustache_token_section_t *token)
+{
+  RendererMustacheData *data = userdata;
+  g_autoptr(GVariant) value = g_variant_dict_lookup_value (data->variables,
+                                                           token->name,
+                                                           NULL);
+
+  if (value == NULL)
+    {
+      g_autofree char *msg = g_strdup_printf ("No such section %s", token->name);
+      (*api->error) (api, data, __LINE__, msg);
+      return 0;
+    }
+
+  if (!_renderer_section_from_ht_variant (value, api, userdata, token))
+    return 0;
+
+  return 1;
+}
+
+static void
+_renderer_set_error (G_GNUC_UNUSED mustache_api_t *api,
+                     void                         *userdata,
+                     G_GNUC_UNUSED uintmax_t       lineno,
+                     const char                   *msg)
+{
+  RendererMustacheData *data = userdata;
+
+  g_set_error (data->error,
+               EKNR_ERROR,
+               EKNR_ERROR_SUBSTITUTION_FAILED,
+               "Failed to perform template substitution: %s",
+               msg);
+}
+
+static mustache_api_t _renderer_mustache_data_vfuncs = {
+  .read = &_renderer_read_from_closure,
+  .write = &_renderer_write_to_closure,
+  .varget = &_renderer_var_from_ht,
+  .sectget = &_renderer_sect_from_ht,
+  .error = &_renderer_set_error
+};
+
+static char *
+_renderer_render_mustache_document_internal (mustache_template_t  *tmpl,
+                                             GVariant             *variables,
+                                             GError              **error)
+{
+  g_autoptr(GVariantDict) variables_dict = g_variant_dict_new (variables);
+  g_autoptr(RendererMustacheData) data = renderer_mustache_data_new (variables_dict,
+                                                                     error,
+                                                                     NULL);
+
+  if (!mustache_render (&_renderer_mustache_data_vfuncs, (void *) data, tmpl))
+    return NULL;
+
+  return g_steal_pointer (&data->output.string);
```

Comment:
This seems like it wouldn't do what it was supposed to. It's `data` that would be autofreed, not its members. And if you aren't returning `data` then where does it get freed?

Reply:
`data` is an autoptr and is cleaned up by `renderer_mustache_data_free`. We steal the actual output string away from it and set it to NULL, so when `renderer_mustache_data_free` is run, it calls `g_clear_pointer` on it, which would be a no-op.

Reply:
I still don't understand this. I don't think you can `g_steal_pointer()` only one member of a struct autoptr.

Reply:
You can in this case - see `renderer_mustache_data_free`. That function will check the member if it is `NULL` before it attempts to free it because it uses `g_clear_pointer`.